File: skills/install-slop-guard/assets/python/rules/smells/no_duplicate_code.py
"""Reject functions with identical bodies."""

from __future__ import annotations

import ast

from ...diagnostics import Diagnostic
from ...rule import Rule

DEFAULT_MIN_STATEMENTS = 3


class NoDuplicateCode(Rule):
    name = "no-duplicate-code"
    group = "smells"
    reference = "https://refactoring.guru/smells/duplicate-code"
    description = (
        "Disallow two functions in a module with identical bodies of `min_statements` statements "
        "or more, 3 by default."
    )

    def run(self) -> list[Diagnostic]:
        minimum = self.threshold("min_statements", DEFAULT_MIN_STATEMENTS)
        seen: dict[str, str] = {}
        for node in ast.walk(self.source.tree):
            if not isinstance(node, (ast.FunctionDef, ast.AsyncFunctionDef)):
                continue
            body = [statement for statement in node.body if not self._is_docstring(statement)]
            if len(body) < minimum:
                continue
            fingerprint = "".join(ast.dump(statement, annotate_fields=False) for statement in body)
            earlier = seen.get(fingerprint)
            if earlier is None:
                seen[fingerprint] = node.name
                continue
            self.report(
                node,
                f"`{node.name}` has the same body as `{earlier}`, so a fix to one has to be "
                f"remembered for the other. Keep one copy and call it from both.",
            )
        return self.diagnostics

    @staticmethod
    def _is_docstring(statement: ast.stmt) -> bool:
        return (
            isinstance(statement, ast.Expr)
            and isinstance(statement.value, ast.Constant)
            and isinstance(statement.value.value, str)
        )

File: skills/install-slop-guard/assets/typescript/smells/rules/no-duplicate-code.ts
import { defineRule } from "@oxlint/plugins";
import type { ESTree } from "@oxlint/plugins";

import { numberOption, thresholdSchema } from "../shared/options.ts";

const DEFAULT_MIN_STATEMENTS = 3;

type FunctionNode = ESTree.ArrowFunctionExpression | ESTree.Function;

function normalise(text: string): string {
	return text.replace(/\s+/gu, " ").trim();
}

/** Ban two functions in one file with identical bodies. */
export const noDuplicateCodeRule = defineRule({
	meta: {
		type: "problem",
		docs: {
			description:
				"Disallow two functions in a file with identical bodies of `minStatements` statements or more.",
		},
		messages: {
			duplicateCode:
				"This function has the same body as an earlier one, so a fix to one has to be remembered for the other. Keep one copy and call it from both.",
		},
		...thresholdSchema({ minStatements: DEFAULT_MIN_STATEMENTS }),
	},
	createOnce(context) {
		const seen = new Set<string>();

		const check = (node: FunctionNode) => {
			const body = node.body;
			if (body === null || body === undefined || body.type !== "BlockStatement") return;
			const minimum = numberOption(context.options, "minStatements", DEFAULT_MIN_STATEMENTS);
			if (body.body.length < minimum) return;
			const fingerprint = normalise(context.sourceCode.getText(body));
			if (!seen.has(fingerprint)) {
				seen.add(fingerprint);
				return;
			}
			context.report({ node, messageId: "duplicateCode" });
		};

		return {
			Program() {
				seen.clear();
			},
			ArrowFunctionExpression: check,
			FunctionDeclaration: check,
			FunctionExpression: check,
		};
	},
});
